hop/io.py

```diff
@@ -3,44 +3,218 @@
 
         Args:
-          broker_url: sets the broker URL to connect to
+            url: Sets the broker URL to connect to.
 
         Kwargs:
-          mode: read ('r') or write ('w') from the stream
-          format: the message serialization format
-          start_at: the message offset to start at
-          config: librdkafka style options, either a `dict` or a file path
+            mode: Read ('r') or write ('w') from the stream.
+            auth: An Auth instance.
+            start_at: The message offset to start at (read only).
+            persist: Whether to listen to new messages forever or stop
+                 when EOS is received (read only).
+            metadata: Whether to receive message metadata along
+                with payload (read only).
+
+        Returns:
+            An open connection to the client, either an adc Producer instance
+            in write mode or an adc Consumer instance in read mode.
+
+        Raises:
+            ValueError: If the mode is not set to read/write or if more than
+                one topic is specified in write mode.
+
+        """
+        group_id, broker_addresses, topics = kafka.parse_kafka_url(url)
+        logger.debug("connecting to addresses=%s  group_id=%s  topics=%s",
+                     broker_addresses, group_id, topics)
+        if not auth and self.auth:
+            auth = self.auth
+        if mode == "w":
+            if len(topics) != 1:
+                raise ValueError("must specify exactly one topic in write mode")
+            if group_id is not None:
+                warnings.warn("group ID has no effect when opening a stream in write mode")
+            if start_at is not None:
+                warnings.warn("start_at has no effect when opening a stream in write mode")
+            if persist is not None:
+                warnings.warn("read_forever has no effect when opening a stream in write mode")
+            return _open_producer(broker_addresses, topics[0], auth=auth)
+        elif mode == "r":
+            if group_id is None:
+                group_id = _generate_group_id(10)
+                logger.info("group ID not specified, generating a random group ID")
+            # set up extra options if provided
+            opts = {}
+            if start_at or self.start_at:
+                opts["start_at"] = start_at if start_at else self.start_at
+            if persist is not None or self.persist is not None:
+                opts["read_forever"] = persist if persist is not None else self.persist
+            return _open_consumer(
+                group_id,
+                broker_addresses,
+                topics,
+                auth=auth,
+                metadata=metadata,
+                **opts,
+            )
+        else:
+            raise ValueError("mode must be either 'w' or 'r'")
+
+
+class Deserializer(Enum):
+    CIRCULAR = models.GCNCircular
+    VOEVENT = models.VOEvent
+    BLOB = models.MessageBlob
```

Comment:
I'm really keen on the plugin approach and was thinking about that myself as well. Others could define new message types and as long as it's registered and defines the right behavior, it can be handled in the same way as the base classes. I think it may be worth it for the reasons you mentioned, despite the extra complexity.

I've been personally eyeing the pluggy library (https://pluggy.readthedocs.io/en/latest/) for supporting a plugin architecture for another project and was wondering your thoughts on it versus rolling our own implementation. It's pretty minimal (1 dependency if python < 3.8) and used to be part of pytest before they split it off into their own thing.

Reply:
Oh, pluggy looks really nice. That would make it so you can just `pip install <some-serialization-plugin>`, which is really simple. I like it.

Reply:
Awesome. We can start planning this out in the near future for users to extend our current classes in a nice way.

Reply:
Made https://github.com/scimma/hop-client/issues/84 to track